tools/drsnoop.py

```diff
@@ -0,0 +1,200 @@
+# latency.
+#
+# Copyright (c) 2019 Ethercflow
+# Licensed under the Apache License, Version 2.0 (the "License")
+#
+# 20-Feb-2019   Ethercflow   Created this.
+
+from __future__ import print_function
+from bcc import ArgString, BPF
+import argparse
+from datetime import datetime, timedelta
+
+# arguments
+examples = """examples:
+    ./drsnoop           # trace all direct reclaim
+    ./drsnoop -T        # include timestamps
+    ./drsnoop -U        # include UID
+    ./drsnoop -P 181    # only trace PID 181
+    ./drsnoop -t 123    # only trace TID 123
+    ./drsnoop -u 1000   # only trace UID 1000
+    ./drsnoop -d 10     # trace for 10 seconds only
+    ./drsnoop -n main   # only print process names containing "main"
+"""
+parser = argparse.ArgumentParser(
+    description="Trace direct reclaim",
+    formatter_class=argparse.RawDescriptionHelpFormatter,
+    epilog=examples)
+parser.add_argument("-T", "--timestamp", action="store_true",
+                    help="include timestamp on output")
+parser.add_argument("-U", "--print-uid", action="store_true",
+                    help="print UID column")
+parser.add_argument("-p", "--pid",
+                    help="trace this PID only")
+parser.add_argument("-t", "--tid",
+                    help="trace this TID only")
+parser.add_argument("-u", "--uid",
+                    help="trace this UID only")
+parser.add_argument("-d", "--duration",
+                    help="total duration of trace in seconds")
+parser.add_argument("-n", "--name",
+                    type=ArgString,
+                    help="only print process names containing this name")
+parser.add_argument("--ebpf", action="store_true",
+                    help=argparse.SUPPRESS)
+args = parser.parse_args()
+debug = 0
+if args.duration:
+    args.duration = timedelta(seconds=int(args.duration))
+
+# load BPF program
+bpf_text = """
+#include <uapi/linux/ptrace.h>
+#include <linux/sched.h>
+
+struct val_t {
+    u64 id;
+    u64 ts; // start time
+    char name[TASK_COMM_LEN];
+};
+
+struct data_t {
+    u64 id;
+    u32 uid;
+    u64 nr_reclaimed;
+    u64 delta;
+    u64 ts;    // end time
+    char name[TASK_COMM_LEN];
+};
+
+BPF_HASH(start, u64, struct val_t);
+BPF_PERF_OUTPUT(events);
+
+int trace_direct_reclaim_begin(void)
+{
+    struct val_t val = {};
+    u64 id = bpf_get_current_pid_tgid();
+    u32 pid = id >> 32; // PID is higher part
+    u32 tid = id;       // Cast and get the lower part
+    u32 uid = bpf_get_current_uid_gid();
+    u64 ts;
+
+    PID_TID_FILTER
+    UID_FILTER
+    if (bpf_get_current_comm(&val.name, sizeof(val.name)) == 0) {
+        val.id = id;
+        val.ts = bpf_ktime_get_ns();
+        start.update(&id, &val);
+    }
+    return 0;
+}
+
+struct reclaimed_args {
+    // from /sys/kernel/debug/tracing/events/vmscan/mm_vmscan_direct_reclaim_-
+    // end/format
+    u64 __unused__;
+    u64 nr_reclaimed;
+};
+
+int trace_direct_reclaim_end(struct reclaimed_args *args)
```

Comment:
could you usee `TRACEPOINT_PROBE` instead of explicit func name `trace_direct_reclaim_end`? The structure `reclaimed_args` also not needed.

Reply:
refactored, PTAL.